Add TorchDispatch version of functionalization

bdhirsh opened the pull request:
This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #105236
* #104974
* #104483
* #106406
* #109023
* #108654
* #108656
* #109248
* #109024
* #108897
* #107917
* __->__ #106404

bdhirsh committed:
Add TorchDispatch version of functionalization

[ghstack-poisoned]

ezyang commented on torch/csrc/autograd/python_torch_functions_manual.cpp:
```diff
@@ -484,6 +484,26 @@ static PyObject* THPVariable__enable_functionalization(
+  if (reapply_views) {
+    at::functionalization::impl::setFunctionalizationReapplyViewsTLS(true);
+  } else {
+    at::functionalization::impl::setFunctionalizationReapplyViewsTLS(false);
```
Don't you already have a binding THPVariable__enable_functionalization for this

bdhirsh replied:
I explicitly added this API because `THPVariable__enable_functionalization` does two things:

(1) sets `reapply_views` TLS to true/false

(2) puts the `Functionalize` key into the dispatcher TLS

And in my usage of functional tensor above, I'm using the fancy new context manager `_IncludeDispatchKeyGuard` to do (1). So I figured I'd separate out these API's.

Some alternatives:

(a) I could also turn this into a python context manager

(b) I could have a dedicated context manager for both (1) and (2), since they're usually used together

I'll probably do one of the two options above, but happy to do either

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,128 @@
+        # - If we want to use _make_subclass(), we have a problem: the subclass will share a TensorImpl with the inner tensor,
+        #   which is of type FunctionalTensorWrapper! We explicitly do not want our wrapper to be a FunctionalTensorWrapper.
+        # - If we use the default tensor.__new__(), we have another problem: it returns inner_tensor.alias(),
+        #   which causes every subclass created above autograd to have autograd view metadata (in addition to also being a FunctionalTensorWrapper).
```
no problem, s'fine

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,128 @@
+        ):
+            try:
+                # By default for python functionalization (for AOTAutograd), we reapply views.
+                torch._functionalize_enable_reapply_views(True)
```
so this no-ops if functionalization was already included, right?

ezyang replied:
unless TLS reentrant rears its head, which I can't easily tell 🤔

bdhirsh replied:
Mentioned below - this doesn't actually toggle the `Functionalize` dispatch key TLS.

But I am in the `IncludeDispatchKeyGuard` context manager above (which will indeed nop if it was already included).

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh commented:
Updated this PR to consolidate all `FunctionalTensor` / `FunctionalTensorMode` changes into a single PR (and separate everything out into other PR's).

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+        yield
+    finally:
+        if maybe_func_mode is not None:
+            torch._C._set_dispatch_mode(maybe_func_mode)
```
In the other PR, you don't use this context manager, instead you enable/disable the mode without removing it from the stack. Is there are a particular reason to do it that way?

bdhirsh replied:
nope - in retrospect, an explicit enable/disable (by adding an extra `is_active` field on the mode) feels wrong. We already have an API now for setting/unsetting the mode, so using that here feels like the right thing.

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+    This class is a lightweight python shim around the C++ functionalization logic.
+
+    FunctionalTensor is required to be used with a corresponding FunctionalTensorMode active.
+    It doesn't bother defining a __torch_dispatch__, because it relies
```
But you do define one lol

bdhirsh replied:
out of date, will fix

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+        unrecognized_types = [
+            t
+            for t in types
+            if not issubclass(t, torch._subclasses.FakeTensor)
```
When do you want to allow subclasses of FakeTensor?

bdhirsh replied:
Yeah you're right, this was probably too preemptive. Will remove.

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+            def unwrap(x):
+                return x.elem
+
+            args_unwrapped = pytree.tree_map_only(FunctionalTensor, unwrap, args)
```
All of these only take a single Tensor argument at the first arg, so it may be marginally faster (compile time wise) to skip the tree map

bdhirsh replied:
good point

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+        )
+
+    def __repr__(self):
+        return f"FunctionalTensor({str(self.elem)})"
```
not repr here?

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+
+        torch._functionalize_enable_reapply_views(True)
+        if x.requires_grad and not x.is_leaf:
+            out = _ToFunctionalTensor.apply(x_functional)
```
This is plausible, but it wasn't immediately obvious to me why you needed this custom autograd function

bdhirsh replied:
Yeah this makes sense - I think I originally wrote the autograd.Function before I'd added the new `mirror_autograd_metadata` method.

Ill try consolidating on `mirror_autograd_metadata`

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+        #   this is handled by FunctionalTensor.to_functional
+        x_functional = torch._to_functional_tensor(x, mirror_autograd_meta=True)
+
+        torch._functionalize_enable_reapply_views(True)
```
This seems suspect. Why are you mutating global state in this function?

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+
+class FunctionalTensorMode(TorchDispatchMode):
+    def __init__(self):
+        self.is_active = True
```
Now that we have a slot for these mode keys, my general attitude is that we shouldn't need `is_active` bool flag. You should explain why `is_active` is not equivalent to just removing the mode from its slot entirely

bdhirsh replied:
You're right - I'll just unset the mode when necessary, thanks!

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+    def __exit__(self, a, b, c):
+        is_on_stack = self.enter_stack.pop()
+        if is_on_stack:
+            super().__exit__(a, b, c)
```
The adjusted semantics I suggested for FakeTensorMode should apply here; doubly so because there are far fewer preexisting uses of FunctionalTensorMode here

ezyang commented on torch/_subclasses/functional_tensor.py:
```diff
@@ -0,0 +1,259 @@
+        ), torch._C._IncludeDispatchKeyGuard(torch._C.DispatchKey.Functionalize):
+            try:
+                # By default for python functionalization (for AOTAutograd), we reapply views.
+                torch._functionalize_enable_reapply_views(True)
```
This should be scoped right? You need to reset it to its old value when you exit this.

ezyang reviewed (approved):
Nice work. This all seems to work out the way I hoped it would.

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]

bdhirsh committed:
Update on "Add TorchDispatch version of functionalization"

This PR adds a new `FunctionalTensor` subclass, and `FunctionalTensorMode` torch dispatch mode. Together, this class/mode are a lightweight wrapper around our existing C++ functionalization logic.

This idea came from Ed - later in the stack, I want to be able to run functionalization **underneath** torch_dispatch, when performing tracing in AOTAutograd. I can't do this easily with vanilla C++ functionalization, because it has a dedicated dispatch key that always runs before TorchDispatch. However, by adding a torch_dispatch mode shim around functionalization, we can use functionalization as a torch_dispatch mode, which will make it easier to run underneath other modes later.

This PR provides the basic new classes, and some light testing.




[ghstack-poisoned]
